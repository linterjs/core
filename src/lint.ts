import { NoLintersError } from "./errors";
import { LinterAdapter } from "./linter-adapter";

export interface LintInput {
  filePath?: string;
  text: string;
}

export const enum LintSeverity {
  WARNING = 1,
  ERROR = 2
}

export interface LintMessage {
  column: number;
  endColumn?: number;
  endLine?: number;
  line: number;
  message: string;
  ruleId: string;
  ruleUri?: string;
  severity: LintSeverity;
}

export interface LintOutput {
  errorCount: number;
  filePath?: string;
  messages: LintMessage[];
  warningCount: number;
}

export type LintFunction = (
  { filePath, text }: LintInput
) => Promise<LintOutput>[];

export function createLint(
  linterAdapterPromises: Set<Promise<LinterAdapter>>
): LintFunction {
  return function lint({ filePath, text }: LintInput): Promise<LintOutput>[] {
    if (linterAdapterPromises.size === 0) {
      throw new NoLintersError();
    }

    const lintArgs = { filePath, text };
    const lintOutput: Promise<LintOutput>[] = [];
    // Run text through all linters
    for (const linterAdapterPromise of linterAdapterPromises.values()) {
      // TODO: Handle error in linterFactory and lint?
      lintOutput.push(
        Promise.resolve(linterAdapterPromise).then(({ lint }) => lint(lintArgs))
      );
    }

    // XXX: Could we return a streaming result for the cli if we lint in parallel?
    // This way the cli could add the linter results for each file as they become
    // available.
    return lintOutput;
  };
}
