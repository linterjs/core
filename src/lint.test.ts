import { createLint } from "./lint";

jest.mock("@linter/eslint");
jest.mock("@linter/prettier");

describe("Lint", () => {
  let lint;
  let linterAdapterPromiseList;
  const text = 'const foo = "bar"';

  test("with no registered linters", () => {
    lint = createLint(new Set());
    expect(() => lint({ text })).toThrowErrorMatchingSnapshot();
  });

  describe("with @linter/eslint", () => {
    beforeAll(() => {
      const { linterProvider: { linterFactory } } = require("@linter/eslint");
      linterAdapterPromiseList = [Promise.resolve(linterFactory())];
      lint = createLint(new Set(linterAdapterPromiseList));
    });

    test("text only", async () => {
      const { linter } = require("@linter/eslint");
      const args = { text };
      const result = await lint(args);
      expect(result).toMatchSnapshot();
      expect(linter.lint).toHaveBeenCalledTimes(1);
      expect(linter.lint).toHaveBeenCalledWith(args);
    });
  });
});
