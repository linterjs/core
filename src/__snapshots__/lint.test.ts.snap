// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lint with @linter/eslint text only 1`] = `
Array [
  Promise {},
]
`;

exports[`Lint with no registered linters 1`] = `"No linters registered. Please install at least one linter provider package."`;
